Add logging to MDANSE and general refactoring

ChiCheng45 opened the pull request:
**Description of work**

- Replaced all print and icecream calls with calls to logging - icecream dependency has now been removed.
- Added a new option in the outputfiles configurations so that they can now output log files.
- Remove unused imports.
- Updated some remaining file headers with the current version.
- Added logs textbox in the running jobs. This outputs INFO level or higher. Warning and errors are coloured.
- Generalised the TriplePanel to MultiPanel
- Updated checkablecombobox in the file selection widget. This is now the same as the checkablecombobox in the atom selection widget.

![image](https://github.com/ISISNeutronMuon/MDANSE/assets/14276033/77c5c90f-22dd-4a21-be38-297171e5c72e)

![image](https://github.com/ISISNeutronMuon/MDANSE/assets/14276033/ab8c1ec5-41a8-46ab-9d68-403c0ac81ce1)

![image](https://github.com/ISISNeutronMuon/MDANSE/assets/14276033/61a09bcc-e895-4293-ba68-b599655d533a)


**To test**
Run a converter job with the log level selected. A log file should be outputted in the same directory as the MDT file but with a .log extension. Do the same with an analysis job. Some jobs might not output any logs. ASE and DOS produce some logs. Check that the logs are updated in the running tab. 

Try rerunning the jobs with "no logs" selected, no logs should be saved.

Also, check with the multicore setting.

ChiCheng45 committed:
updated triple panel to a multi panel

ChiCheng45 committed:
corrected multipanel

ChiCheng45 committed:
job logging info panel initial commit

ChiCheng45 committed:
print to log in Ijob

ChiCheng45 committed:
refactored code and fix logging for multicore jobs

ChiCheng45 committed:
added process to logging format string

ChiCheng45 committed:
applied black

ChiCheng45 committed:
Merge branch 'protos' into chi/logger

ChiCheng45 committed:
applied black

ChiCheng45 committed:
convert print statements to log calls

ChiCheng45 committed:
convert icecream calls to log calls and cleanedup imports

ChiCheng45 committed:
corrected file headers

ChiCheng45 committed:
import fix

ChiCheng45 committed:
import fix

ChiCheng45 committed:
add streamhandler to logger when running GUI, updated listener to add all handler in the logger and change log level in hdftrajinputdata

ChiCheng45 committed:
applied black

ChiCheng45 committed:
refactored code

ChiCheng45 committed:
updated so that logs are color when they are warning, error or critical messages

ChiCheng45 committed:
applied black

ChiCheng45 committed:
updated outputfile widget to use atom selections checkablecombobox

ChiCheng45 committed:
updated type hints

ChiCheng45 committed:
updated file dialog labels

ChiCheng45 committed:
outputfiles widgets updates when combobox selection changes. Fix for no formats selected.

ChiCheng45 committed:
Updated so that all jobs and converters write logs to a logfile when requested and fixed tests.

ChiCheng45 committed:
output files logs gui fix

ChiCheng45 committed:
updated tests to test for log output

ChiCheng45 committed:
unit tests fixes

ChiCheng45 committed:
fix logging with multiprocessing

ChiCheng45 committed:
edit

ChiCheng45 committed:
converter test fixes

ChiCheng45 committed:
updated so that log level can be chosen

ChiCheng45 committed:
applied black

MBartkowiakSTFC commented:
This is a big improvement throughout. Thank you for cleaning up the code and improving the GUI components. The MultiPanel layout and single implementation of the checkable combo box make the code more consistent.

One comment: you already added a method (add_log_file_handler) to IJob, which enables logging. Do you think that it would be possible to add another, which will wrap up the entire
```
        if self.configuration["output_files"]["write_logs"]:
            log_filename = self.configuration["output_files"]["root"] + ".log"
            self.add_log_file_handler(
                log_filename, self.configuration["output_files"]["log_level"]
            )
```
sequence, so we avoid repeating the same code? You could just have one method in IJob which checks which one of configuration["output_files"] or configuration["output_file"] is present, or have the Converter implementation of the method replace the "output_files" with "output_file". What do you think?

Otherwise everything is working as expected. Two things may have to be added **later**:
1. We will need to ensure that every job that fails outputs an error message. This is not guaranteed at the moment.
2. The GUI log viewer could benefit from an option of setting the logging level it should display. If we decide to do it, it should only happen after we have merged the changes to the settings files.

ChiCheng45 committed:
changed so that all jobs outputfiles have "output_files" key. Updated so that the initialize method in IJob add the log file handler. All jobs initialize methods call super.

ChiCheng45 committed:
test fixes

ChiCheng45 commented:
Thanks for the suggestions. I've added the updated IJobs so it checks for the "output_files" key and adds the file handler.

```
    def initialize(self):
        if (
            "output_files" in self.configuration
            and self.configuration["output_files"]["write_logs"]
        ):
            log_filename = self.configuration["output_files"]["root"] + ".log"
            self.add_log_file_handler(
                log_filename, self.configuration["output_files"]["log_level"]
            )
```

all jobs now call `super().initialize()`. I've also replaced the converter settings keyword from "output_file" with "output_files".


And yes I agree both points (1) and (2) should be added in the future. For (2) we could put the GUI log level chosen in the settings.

MBartkowiakSTFC reviewed (approved):
Thank you, I think this PR is ready to be merged now.
